File: Chapter 1/SpeedTest/objc/main.m
//
//  main.m
//  objc
//
//  Created by Kostiantyn Koval on 01/08/15.
//  Copyright © 2015 Kostiantyn Koval. All rights reserved.
//

#import <Foundation/Foundation.h>
@import QuartzCore;

const NSUInteger MAX_COUNT = 100000;
const NSUInteger NUM_TRIALS = 20;

NSMutableArray* makeArray(NSUInteger count) {
  NSMutableArray *array = [[NSMutableArray alloc] initWithCapacity:count];
  
  for (NSUInteger i = 0; i < count; i++) {
    NSNumber *number = @(arc4random_uniform(UINT32_MAX));
    [array addObject:number];
  }
  return array;
}


int main(int argc, const char * argv[]) {
  @autoreleasepool {
    
    double sumTime = 0;
    for (int i = 1; i <= NUM_TRIALS; i++) {
      NSMutableArray *unsortedArray = makeArray(MAX_COUNT);
      
      CFTimeInterval startTime = CACurrentMediaTime();
      [unsortedArray sortUsingComparator:^NSComparisonResult(NSNumber *n1, NSNumber *n2) {
        return [n1 compare:n2];
      }];
      CFTimeInterval totalTime = CACurrentMediaTime() - startTime;
      NSLog(@"time: %g s", totalTime);
      
      sumTime += totalTime;
    }
    NSLog(@"RESULT: %g s", sumTime / NUM_TRIALS);
  }
  return 0;
}

File: Chapter 1/SpeedTest/c/main.c
//
//  main.c
//  c
//
//  Created by Kostiantyn Koval on 01/08/15.
//  Copyright © 2015 Kostiantyn Koval. All rights reserved.
//

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

static const uint32_t MAX_COUNT = 100000;
static const uint32_t NUM_TRIALS = 20;

int compare(const void * a, const void * b)
{
  
  if ( *(uint32_t*)a < *(uint32_t*)b ) {
    return -1;
  }
  
  if ( *(uint32_t*)a > *(uint32_t*)b ) {
    return 1;
  }
  
  return 0;
}

void makeArray(uint32_t arr[], const unsigned int count)
{
  for (unsigned int i = 0; i < count; i++) {
    arr[i] = arc4random_uniform(UINT32_MAX);
  }
}

int main(int argc, const char * argv[]) {
  
  uint32_t array[MAX_COUNT];
  double resultsSum = 0;
  
  for (unsigned int i = 0; i < NUM_TRIALS; i++) {
    
    makeArray(array, MAX_COUNT);
    
    double startTime = clock();
    qsort(array, MAX_COUNT, sizeof(uint32_t), compare);
    double endTime = clock();
    
    double totalTime = (endTime - startTime) / CLOCKS_PER_SEC;
    resultsSum += totalTime;
    printf("time: %.6lf s\n", totalTime);
    
    memset(&array[0], 0, sizeof(array));
  }
  printf("\nRESULT: %.6lf s", resultsSum / NUM_TRIALS);
  return 0;
}
